### ISSUE TITLE
Supporting multiple sets of proto compilers for the same proto file

### ISSUE THREAD

**Issue by linzhp** [bazelbuild/bazel-gazelle#497]
When there are many teams sharing the same proto file, we run into cases when team A wants the proto file be compiled with `compilers = ['a', 'b']` and team B wants it to be compiled with `compilers = ['c', 'd']` in `go_proto_library`. In this case, we need two `go_proto_library` rules sharing the same `proto_library`.

We have the same requirement for thrift too. But since thrift is not supported in Gazelle, we implemented it as Gazelle extensions internally. What we did was to give a name to each set of compilers, and design a `# gazelle:thrift_compilers x,y` directive to tell the extension to generate a go_thrift_library rule with compiler set `x`, which is `[a, b]` and another one with `y`, which is `[c, d]`. The `importpath` for these go_thrift_library rule has a different prefix. Each compiler set has its own Go import prefix.

We are planning to do the same for proto. Now we have two options:

* Light approach. Add another Gazelle directive to Gazelle's Go extension, so we can enable Gazelle's official proto extension to generate `proto_library` rules as usual while stop Go extension from generating `go_proto_library`. Then we implement a goproto extension with similar design as our own gothrift extension.
* Heavy approach. Refactor Gazelle's Go extension, possibly move it out of Go extension, and support different set of compilers with different Go prefix

I feel the light approach is less disruptive, as mapping an import prefix to a set of compilers is a quite different design than that of current official Gazelle extensions, which maps an import prefix to a directory.

@jayconrod What do you think?

**Comment by jayconrod** [comment]
The light approach makes the most sense to me. This is a pretty customized situation, and it would be hard to make the general Go extension support this in a way that would be useful to everyone without adding a lot of complexity.

So maybe the new directive is `# gazelle:go_proto <mode>` where `<mode>` is `default` (current behavior) or `disable` (don't generate `go_proto_library`)? Hopefully more modes aren't needed in the future, but it feels like we need more than just a boolean here.

**Comment by linzhp** [comment]
Sounds good. One of us may come up with a diff adding `go_proto` later.

**Comment by achew22** [comment]
One of the things I've been considering is trying to add a mapping that says "if you're importing this `.proto` file then you will also need to add this compiler". That way you could have something like

```
# gazelle:go_extra_compilers google/api/annotations.proto @com_github_grpc_ecosystem_grpc_gateway//protoc-gen-grpc-gateway:go_gen_grpc_gateway
```

which would detect when you `import "google/api/annotations.proto"` in your code and add `@com_github_grpc_ecosystem_grpc_gateway//protoc-gen-grpc-gateway:go_gen_grpc_gateway` as a compiler to the go_proto_library.

WDYT?

**Comment by linzhp** [comment]
The main issue we run into is some proto files need to be compiled by multiple and incompatible sets of compilers. So we need multiple go_proto_library rules for the same proto_library. Not sure if your directive can solve this

**Comment by linzhp** [comment]
Resolved by #530
